components/measure-upload/MeasureRepositoryUpload.tsx

```diff
@@ -0,0 +1,235 @@
+  const [isLoadingPackage, setIsLoadingPackage] = useState<PACKAGE_STATES>(PACKAGE_STATES.NONE);
+  const [measureSelectIsOpen, setMeasureSelectIsOpen] = useState(false);
+
+  const [{ measureRepositoryUrl, isFile, selectedMeasureId, displayMap }, setMeasureBundle] =
+    useRecoilState(measureBundleState);
+  const setMeasurementPeriod = useSetRecoilState(measurementPeriodState);
+
+  const idSelectData = useRecoilValue(displayMapToSelectDataState);
+
+  useEffect(() => {
+    if (isFile) {
+      setMeasureSelectIsOpen(false);
+      setIsLoadingPackage(PACKAGE_STATES.NONE);
+    }
+  }, [isFile]);
+
+  const rejectUpload = (
+    message: string | string[],
+    attemptedBundleDisplay: string | null,
+    isValueSetMissingError = false
+  ) => {
+    logError({
+      id: uuidv4(),
+      message,
+      timestamp: new Date().toISOString(),
+      attemptedBundleDisplay,
+      isValueSetMissingError
+    });
+    showNotification({
+      icon: <IconAlertCircle />,
+      title: 'Bundle upload failed',
+      message: 'There was an issue uploading your bundle. Please correct the issues listed below.',
+      color: 'red'
+    });
+    setIsLoadingPackage(PACKAGE_STATES.FAIL);
+  };
+
+  /*
+    Handles updating of the Measure Repository Service url field and associated state variables
+  */
+  function handleMrsUrlChange(event: React.ChangeEvent<HTMLInputElement>) {
+    if (measureSelectIsOpen) {
+      setMeasureSelectIsOpen(false);
+    }
+    setMeasureBundle(mb => ({
+      ...mb,
+      measureRepositoryUrl: event.currentTarget.value,
+      selectedMeasureId: null,
+      displayMap: {},
+      ...(!mb.isFile ? { content: null } : undefined)
+    }));
+    setIsLoadingPackage(PACKAGE_STATES.NONE);
+  }
+
+  /*
+    Takes the entries from a Measure retrieval response and creates a map from Measure id
+    to Measure display label for Measure select component
+  */
+  function createMeasureDisplayMap(
+    displayMap: Record<string, string>,
+    { resource: measure }: { resource: fhir4.Measure }
+  ) {
+    let label: string;
+    const officialIdentifier = measure.identifier?.find(
+      identifier =>
+        identifier.use === 'official' && identifier.system === 'http://hl7.org/fhir/cqi/ecqm/Measure/Identifier/cms'
+    );
+    if (officialIdentifier?.value) {
+      label = measure.version ? `${officialIdentifier.value}|${measure.version}` : officialIdentifier.value;
+    } else if (measure.name) {
+      label = measure.version ? `${measure.name}|${measure.version}` : measure.name;
+    } else {
+      label = measure.id as string;
+    }
+    displayMap[measure.id as string] = label;
+    return displayMap;
+  }
+
+  /*
+    Sends a `GET` request to the `/Measure` endpoint of the specified Measure Repository Service.
+    Parses and stores the measureIds if successful
+  */
+  async function retrieveMeasures() {
+    setIsLoadingIds(true);
+    try {
+      const response = await fetch(`${measureRepositoryUrl}/Measure`);
```

Comment:
Nitpick here so I'm open to others' thoughts on this... is it worth adding a check to remove a trailing slash from `measureRepositoryUrl` before doing the `fetch`? I don't think we've added this handling in the past to our FHIR apps but I've noticed that a lot of SMART applications check for trailing slash and it makes things a little easier for the user (the "user" being me... I always type in `http://localhost:3000/4_0_1/` by default lol. not sure if others do this as well)

Reply:
Hahaha good call. Seems like it'll save some grief, and easy enough to implement